hare4/hare_test.go

```diff
@@ -947,26 +1121,281 @@ func TestHare_AddProposal(t *testing.T) {
+					if err := other.hare.Handler(ctx, n.peerId(), msg); err != nil {
+						panic(err)
+					}
+				}
+				return nil
+			}).
+			AnyTimes()
+		n.mockStreamRequester.EXPECT().StreamRequest(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Do(
+			func(ctx context.Context, p p2p.Peer, msg []byte, cb server.StreamRequestCallback, _ ...string) error {
+				for _, other := range cluster.nodes {
+					if other.peerId() == p {
+						b := make([]byte, 0, 1024)
+						buf := bytes.NewBuffer(b)
+						if err := other.hare.handleProposalsStream(ctx, msg, buf); err != nil {
+							return fmt.Errorf("exec handleProposalStream: %w", err)
+						}
+						if err := cb(ctx, buf); err != nil {
+							return fmt.Errorf("exec callback: %w", err)
+						}
+					}
+				}
+				return nil
 			},
+		).AnyTimes()
+	}
+
+	cluster.genProposals(layer, 2)
+	cluster.genProposalNode(layer, 2)
+	cluster.movePreround(layer)
+	for i := 0; i < 2*int(notify); i++ {
+		cluster.moveRound()
+	}
+	var consistent []types.ProposalID
+	cluster.waitStopped()
+	for _, n := range cluster.nodes {
+		select {
+		case coin := <-n.hare.Coins():
+			require.Equal(t, coin.Layer, layer)
+		default:
+			require.FailNow(t, "no coin")
 		}
-		require.EqualValues(t, cfg.Committee, cfg.CommitteeFor(0))
-		require.EqualValues(t, cfg.Committee, cfg.CommitteeFor(15))
-		require.EqualValues(t, 50, cfg.CommitteeFor(16))
-		require.EqualValues(t, 50, cfg.CommitteeFor(100))
-	})
+		select {
+		case rst := <-n.hare.Results():
+			require.Equal(t, rst.Layer, layer)
+			require.NotEmpty(t, rst.Proposals)
+			if consistent == nil {
+				consistent = rst.Proposals
+			} else {
+				require.Equal(t, consistent, rst.Proposals)
+			}
+		default:
+			require.FailNow(t, "no result")
+		}
+		require.Empty(t, n.hare.Running())
+	}
+}
+
+// TestHare_ReconstructAll tests that the nodes go into a
+// full message exchange in the case that a signature fails
+// although all compact hashes and proposals match.
+func TestHare_ReconstructAll(t *testing.T) {
+	cfg := DefaultConfig()
+	cfg.LogStats = true
+	tst := &tester{
+		TB:            t,
+		rng:           rand.New(rand.NewSource(1001)),
+		start:         time.Now(),
+		cfg:           cfg,
+		layerDuration: 5 * time.Minute,
+		beacon:        types.Beacon{1, 1, 1, 1},
+		genesis:       types.GetEffectiveGenesis(),
+	}
+	cluster := newLockstepCluster(tst, withMockVerifier()).
+		addActive(3)
+	if cluster.signersCount > 0 {
+		cluster = cluster.addSigner(cluster.signersCount)
+		cluster.partitionSigners()
+	}
+	layer := tst.genesis + 1
+
+	cluster.drainInteractiveMessages()
+	// cluster setup
+	calls := [3]int{}
+	for i, n := range cluster.nodes {
+		n.mverifier.EXPECT().Verify(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+			DoAndReturn(func(_ signing.Domain, _ types.NodeID, _ []byte, _ types.EdSignature) bool {
+				calls[i] = calls[i] + 1
+				// when first call return false, other
+				return !(calls[i] == 1)
+			}).AnyTimes()
+	}
```

Comment:
How about creating proposals with invalid signatures instead of mocking the verifier?

If mocking is easier/better, consider rewriting this as:
```go
	for _, n := range cluster.nodes {
		gomock.InOrder(
			n.mverifier.EXPECT().
				Verify(signing.PROPOSAL, gomock.Any(), gomock.Any(), gomock.Any()).
				Return(false).
				MaxTimes(1),
			n.mverifier.EXPECT().
				Verify(signing.PROPOSAL, gomock.Any(), gomock.Any(), gomock.Any()).
				Return(true).
				AnyTimes(),
		)
	}
```

Reply:
This is a bit more complicated because the mocked `Verify` is not because the proposal signatures are incorrect, but because the node which receives the proposal matches the wrong proposal because it has some collision locally, and then cannot verify the signature correctly.

Reply:
OK, I understand. Why not generate colliding proposals in a way that would trigger the desired flow? I generally prefer to simulate real flow (by creating realistic conditions that trigger the tested behavior) rather than hacking the insides of the implementation. Such tests are easier to understand and maintain.

Reply:
Because those things require mining of the actual data, building external tools, the tests would start breaking in case we change small details inside of the actual domain object. Pluggable implementations are standard practice and personally I'd rather go down that way.